### ISSUE TITLE
`renderHTTP()` calls `http_*` with bad request if `checkPreconditions()` defers.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3084]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)| @wsanchez reported|
|-|-|
|Trac ID|trac#3084|
|Type|defect|
|Created|2008-03-17 04:14:16Z|

`renderHTTP()` calls `http_*` with request set to the result of `checkPreconditions()` (typically `None`) in cases where `checkPreconditions()` returns a deferred.  This is unfortunate in those cases.

<details><summary>Searchable metadata</summary>

```
trac-id__3084 3084
type__defect defect
reporter__wsanchez wsanchez
priority__highest highest
milestone__Web2_Gold_Master Web2-Gold-Master
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__ 
time__1205727256000000 1205727256000000
changetime__1205794000000000 1205794000000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_1"></a>|@wsanchez commented|
|-|-|

(In [22859]) renderHTTP() calls http_* with request set to the result of checkPreconditions() (typically None) in cases where checkPreconditions() returns a deferred. This is unfortunate in those cases.

Add a test case for this, and while I'm at it, tests for a lot of other stuff in resource.py.

Refs [#3084](https://github.com/twisted/twisted/issues/3084).

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_2"></a>|@wsanchez set owner to @exarkun|
|-|-|

See r22859 for proposed fix and test, plus a bunch of bonus tests.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_3"></a>|@wsanchez set owner to @dreid|
|-|-|

The line that specifically tests this bug is 29: ```assert request is not None```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_4"></a>|@jyknight commented|
|-|-|

Looks good to me, other than the use of assert instead of an explicit raise, but I'm probably not an anal enough reviewer Twisted.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_5"></a>|@wsanchez commented|
|-|-|

(In [22910]) Remove assert and return an error instead for trial cleanliness.

Refs [#3084](https://github.com/twisted/twisted/issues/3084).

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_6"></a>|@wsanchez set owner to @wsanchez<br>@wsanchez set status to `assigned`|
|-|-|

r22910 removes the assert.  Checked that the test still catches this bug by unfixing it and verifying that it fails as expected.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_8"></a>|@wsanchez set status to `closed`|
|-|-|

(In [22911]) Merge r22859 and r22910 from branches/dav-take-two-3081-1: Invalid request object from renderHTTP()

renderHTTP() calls http_* with bad request if checkPreconditions() defers.

Author: wsanchez
Reviewer: foom
Fixes [#3084](https://github.com/twisted/twisted/issues/3084)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
